File: apps/android/feature/map/src/main/kotlin/com/sigmundgranaas/turbo/expressive/feature/map/RouteViewModel.kt
package com.sigmundgranaas.turbo.expressive.feature.map

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.sigmundgranaas.turbo.expressive.core.data.FollowController
import com.sigmundgranaas.turbo.expressive.core.data.PathRepository
import com.sigmundgranaas.turbo.expressive.core.data.RouteRepository
import com.sigmundgranaas.turbo.expressive.core.geo.GeoMetrics
import com.sigmundgranaas.turbo.expressive.core.geo.GeoPath
import com.sigmundgranaas.turbo.expressive.core.geo.GeoPathSource
import com.sigmundgranaas.turbo.expressive.core.map.OfflineTileManager
import com.sigmundgranaas.turbo.expressive.core.map.RouteCorridor
import com.sigmundgranaas.turbo.expressive.domain.BaseLayer
import com.sigmundgranaas.turbo.expressive.domain.DownloadSpec
import com.sigmundgranaas.turbo.expressive.domain.LatLng
import com.sigmundgranaas.turbo.expressive.domain.RoutePlan
import com.sigmundgranaas.turbo.expressive.domain.RoutePreset
import com.sigmundgranaas.turbo.expressive.domain.RouteStreamEvent
import com.sigmundgranaas.turbo.expressive.domain.SavedPath
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.util.UUID
import javax.inject.Inject

sealed interface RouteUiState {
    data object Idle : RouteUiState
    data class Solving(val progress: List<LatLng>) : RouteUiState
    data class Done(val plan: RoutePlan) : RouteUiState
    data class Following(val plan: RoutePlan) : RouteUiState
    data class Error(val message: String) : RouteUiState

    /** The polyline to draw for this state (empty when nothing to show). */
    val polyline: List<LatLng>
        get() = when (this) {
            is Solving -> progress
            is Done -> plan.geometry
            is Following -> plan.geometry
            else -> emptyList()
        }
}

/** Drives a route solve against the Turbo pathfinder, streaming live progress. */
@HiltViewModel
class RouteViewModel @Inject constructor(
    private val routes: RouteRepository,
    private val paths: PathRepository,
    private val offline: OfflineTileManager,
    private val follow: FollowController,
) : ViewModel() {
    private val _state = MutableStateFlow<RouteUiState>(RouteUiState.Idle)
    val state: StateFlow<RouteUiState> = _state.asStateFlow()

    private val _preset = MutableStateFlow(RoutePreset.Balanced)
    val preset: StateFlow<RoutePreset> = _preset.asStateFlow()

    /** Ordered route waypoints: first is the origin, last the destination, the rest stops. */
    private val _waypoints = MutableStateFlow<List<LatLng>>(emptyList())
    val waypoints: StateFlow<List<LatLng>> = _waypoints.asStateFlow()

    private var job: Job? = null
    private var resumeFollowing = false
    private val undoStack = ArrayDeque<List<LatLng>>()

    // A waypoint is being dragged on the map. While true, re-solves are DEFERRED so the route
    // line can't thrash under the user's finger (e.g. a follow-mode reroute landing mid-drag);
    // the deferred solve runs on [endWaypointDrag]. The drop's own commit ends the drag first,
    // so the moved waypoint re-solves normally.
    private var waypointDragActive = false
    private var solvePendingAfterDrag = false

    init {
        // Keep the route state in step with the follow session: if following ends from
        // outside (e.g. the lock-screen Stop button → FollowController.stop()), drop back
        // to Idle so the in-app surface and the widget can't disagree.
        viewModelScope.launch {
            follow.session.collect { s ->
                if (!s.active && _state.value is RouteUiState.Following) _state.value = RouteUiState.Idle
            }
        }
    }

    /** Start a fresh two-point route (origin → destination). */
    fun planRoute(from: LatLng, to: LatLng, preset: RoutePreset = _preset.value) {
        _preset.value = preset
        undoStack.clear()
        setWaypoints(listOf(from, to), debounce = false)
    }

    /**
     * Add an intermediate stop, inserted at the least-detour position between
     * existing waypoints, then re-solve. No-op until a route exists.
     */
    fun addStop(point: LatLng) {
        val current = _waypoints.value
        if (current.size < 2) return
        pushUndo()
        setWaypoints(Waypoints.insertLeastDetour(current, point), debounce = true)
    }

    /**
     * Extend the route by appending a waypoint to the END (the new destination; the old
     * destination becomes a via). This is what sequential map taps do — predictable
     * order — as opposed to [addStop], which inserts a POI at the least-detour position.
     */
    fun appendWaypoint(point: LatLng) {
        val current = _waypoints.value
        if (current.size < 2) return
        pushUndo()
        setWaypoints(current + point, debounce = true)
    }

    /** Remove the waypoint at [index] (a route needs at least two; otherwise clears). */
    fun removeWaypoint(index: Int) {
        val current = _waypoints.value
        if (index !in current.indices) return
        pushUndo()
        val next = current.toMutableList().apply { removeAt(index) }
        if (next.size < 2) clear() else setWaypoints(next, debounce = true)
    }

    /** A map drag of a waypoint began — defer re-solves until it ends (see [waypointDragActive]). */
    fun beginWaypointDrag() {
        waypointDragActive = true
    }

    /** The map drag ended — resume re-solving, running any solve deferred during the drag. */
    fun endWaypointDrag() {
        waypointDragActive = false
        if (solvePendingAfterDrag) {
            solvePendingAfterDrag = false
            solve(debounce = true)
        }
    }

    /** Move the waypoint at [index] to a new position (drag on the map), then re-solve. */
    fun moveWaypointTo(index: Int, point: LatLng) {
        val current = _waypoints.value
        if (index !in current.indices) return
        pushUndo()
        setWaypoints(current.toMutableList().also { it[index] = point }, debounce = true)
    }

    /** Reorder a waypoint (drag), then re-solve. */
    fun moveWaypoint(from: Int, to: Int) {
        val current = _waypoints.value
        if (from !in current.indices || to !in current.indices || from == to) return
        pushUndo()
        val next = current.toMutableList().apply { add(to, removeAt(from)) }
        setWaypoints(next, debounce = true)
    }

    /** Revert the last waypoint edit. */
    fun undo() {
        val previous = undoStack.removeLastOrNull() ?: return
        if (previous.size < 2) clear() else setWaypoints(previous, debounce = false)
    }

    val canUndo: Boolean get() = undoStack.isNotEmpty()

    private fun pushUndo() {
        undoStack.addLast(_waypoints.value)
        if (undoStack.size > UNDO_LIMIT) undoStack.removeFirst()
    }

    private fun setWaypoints(points: List<LatLng>, debounce: Boolean) {
        _waypoints.value = points
        solve(debounce)
    }

    private fun solve(debounce: Boolean) {
        val points = _waypoints.value
        if (points.size < 2) return
        // Mid-drag: hold off. The waypoint state is already updated for the live overlay; the
        // re-solve runs once the finger lifts ([endWaypointDrag]) so the line doesn't thrash.
        if (waypointDragActive) {
            solvePendingAfterDrag = true
            return
        }
        job?.cancel()
        val current = _state.value
        // An off-route reroute while FOLLOWING must be INVISIBLE: re-solving in the background
        // and only swapping the route LINE in when the new plan lands. Dropping to Solving
        // here would tear down the live follow sheet, reset the map's bottom inset and un-dim
        // the covered line — a jarring view change mid-navigation. So while following we keep
        // showing Following(oldPlan) for the whole re-solve and ignore intermediate progress.
        // [resumeFollowing] is set by [reroute]; a first-time follow or a stop edit is unaffected.
        val silentFollow = resumeFollowing && current is RouteUiState.Following
        // Graceful re-route for a solved/following route: keep the current line on screen until
        // the new result lands, so the path doesn't snap to a straight line and back. Only the
        // very first solve (no prior geometry) shows the live straight-line→refined progress.
        val previous = when (current) {
            is RouteUiState.Done -> current.plan.geometry
            is RouteUiState.Following -> current.plan.geometry
            else -> null
        }
        if (!silentFollow) _state.value = RouteUiState.Solving(previous ?: points)
        job = viewModelScope.launch {
            if (debounce) kotlinx.coroutines.delay(DEBOUNCE_MS)
            try {
                routes.planStream(points, _preset.value).collect { event ->
                    when (event) {
                        is RouteStreamEvent.Progress ->
                            if (!silentFollow) {
                                _state.value =
                                    if (previous != null) RouteUiState.Solving(previous)
                                    else RouteUiState.Solving(event.coordinates)
                            } // following: stay on the old line, no flicker
                        is RouteStreamEvent.Result ->
                            _state.value =
                                if (resumeFollowing) { resumeFollowing = false; RouteUiState.Following(event.plan) }
                                else RouteUiState.Done(event.plan)
                        is RouteStreamEvent.Failure ->
                            // A failed reroute must not blow away the live view — keep following
                            // the existing line; only a normal solve surfaces the error.
                            if (!silentFollow) _state.value = RouteUiState.Error(event.message) else resumeFollowing = false
                    }
                }
            } catch (t: Throwable) {
                if (!silentFollow) _state.value = RouteUiState.Error("Couldn't reach the router.") else resumeFollowing = false
            }
        }
    }

    /** Re-solve from the current position to the same destination, staying in Follow mode. */
    fun reroute(from: LatLng) {
        val current = _waypoints.value
        if (current.size < 2 || _state.value !is RouteUiState.Following) return
        // One reroute at a time: the off-route check keeps seeing Following (the view never
        // leaves follow mode), so without this guard each new fix while off-route would pile
        // on another re-solve.
        if (resumeFollowing) return
        resumeFollowing = true
        // Replace the origin with the live position, keep the stops + destination.
        setWaypoints(listOf(from) + current.drop(1), debounce = false)
    }

    /** Re-plan the current trip with a different style. */
    fun selectPreset(preset: RoutePreset) {
        _preset.value = preset
        if (_waypoints.value.size >= 2) solve(debounce = false)
    }

    /** Enter turn-by-route following for a solved route. */
    /** Load a saved track by id (for "open on map"). */
    suspend fun pathById(id: String): com.sigmundgranaas.turbo.expressive.domain.SavedPath? = paths.byId(id)

    fun follow(nearbyCheckpoints: List<Pair<LatLng, String>> = emptyList()) {
        (_state.value as? RouteUiState.Done)?.let {
            _state.value = RouteUiState.Following(it.plan)
            // Checkpoints = the stops after the origin (US-3), labelled by their on-map letter
            // (B, C, …) so the split list matches the waypoint badges. Saved markers that sit near
            // the route are folded in as extra checkpoints (D3), then everything is ordered by
            // arc-length along the route so splits fire in travel order regardless of source.
            val stops = _waypoints.value.drop(1)
            val stopCps = stops.mapIndexed { i, p -> p to ('B' + i).toString() }
            val merged = (stopCps + nearbyCheckpoints)
                .sortedBy { (pos, _) -> GeoMetrics.arcLengthAlong(it.plan.geometry, pos) }
            follow.start(
                it.plan,
                phasePoints = merged.map { (pos, _) -> pos },
                phaseNames = merged.map { (_, name) -> name },
            )
        }
    }

    /** The live follow session backing the lock-screen widget; same source the sheet reads. */
    val followSession get() = follow.session

    /** Pause the follow — capture keeps buffering (US-4, Follow = Record). */
    fun pauseFollow() = follow.pause()

    /** Resume a paused follow; [include] stitches the paused-buffer walk onto the track (US-4). */
    fun resumeFollow(include: Boolean) = follow.resume(include)

    /** One-tap pause/resume (resuming this way discards the buffer; the UI prompts a big one). */
    fun toggleFollowPause() = follow.togglePause()

    /**
     * Follow an already-saved track's geometry (no solve) — used by "Follow" on a
     * saved track opened on the map, so following works for recorded/imported lines
     * the same as for planned routes.
     */
    fun followTrack(geometry: List<LatLng>, distanceM: Double, ascentM: Double, durationS: Double, name: String? = null) {
        if (geometry.size < 2) return
        job?.cancel()
        val plan = RoutePlan(
            distanceM = distanceM,
            durationS = durationS,
            ascentM = ascentM,
            onTrailPct = 0.0,
            surfaces = emptyMap(),
            geometry = geometry,
        )
        _state.value = RouteUiState.Following(plan)
        follow.start(plan, name = name)
    }

    fun clear() {
        job?.cancel()
        follow.stop()
        _waypoints.value = emptyList()
        undoStack.clear()
        _state.value = RouteUiState.Idle
    }

    /** Queue an offline download of the padded corridor around the solved route. */
    fun downloadAlongRoute(base: BaseLayer, name: String = "Route area") {
        val geometry = when (val s = _state.value) {
            is RouteUiState.Done -> s.plan.geometry
            is RouteUiState.Following -> s.plan.geometry
            else -> return
        }
        val bounds = RouteCorridor.bounds(geometry) ?: return
        offline.download(DownloadSpec(name = name, base = base, bounds = bounds, minZoom = 8.0, maxZoom = 15.0))
    }

    fun saveAsTrack(name: String, kind: com.sigmundgranaas.turbo.expressive.domain.ActivityKindId? = null) {
        val plan = when (val s = _state.value) {
            is RouteUiState.Done -> s.plan
            is RouteUiState.Following -> s.plan
            else -> return
        }
        if (plan.geometry.size < 2) return
        val geo = GeoPath(
            points = plan.geometry,
            source = GeoPathSource.Route,
            distanceM = plan.distanceM,
            ascentM = plan.ascentM,
            movingTimeSeconds = plan.durationS.toInt(),
            recordedAtEpochMs = System.currentTimeMillis(),
        )
        viewModelScope.launch {
            paths.save(SavedPath(id = "p-${UUID.randomUUID()}", name = name.ifBlank { "Route" }, path = geo, activityKind = kind))
        }
    }

    /**
     * Save a hand-built line (Create-track Line/Draw modes) as a track. Distance
     * is derived from the geometry; no solve, no surfaces.
     */
    fun saveLine(name: String, geometry: List<LatLng>) {
        if (geometry.size < 2) return
        val geo = GeoPath.fromPoints(geometry, GeoPathSource.Measure)
            .copy(recordedAtEpochMs = System.currentTimeMillis())
        viewModelScope.launch {
            paths.save(SavedPath(id = "p-${UUID.randomUUID()}", name = name.ifBlank { "Track" }, path = geo))
        }
    }

    private companion object {
        const val DEBOUNCE_MS = 300L
        const val UNDO_LIMIT = 20
    }
}

/** Pure waypoint geometry, isolated for testability. */
internal object Waypoints {
    /**
     * Insert [point] into [waypoints] at the position that adds the least extra
     * straight-line distance — i.e. on the segment whose detour through the point
     * is smallest. Endpoints (origin, destination) are never displaced.
     */
    fun insertLeastDetour(waypoints: List<LatLng>, point: LatLng): List<LatLng> {
        if (waypoints.size < 2) return waypoints + point
        var bestIndex = 1
        var bestDelta = Double.MAX_VALUE
        for (i in 0 until waypoints.size - 1) {
            val a = waypoints[i]
            val b = waypoints[i + 1]
            val delta = GeoMetrics.haversineMeters(a, point) +
                GeoMetrics.haversineMeters(point, b) -
                GeoMetrics.haversineMeters(a, b)
            if (delta < bestDelta) { bestDelta = delta; bestIndex = i + 1 }
        }
        return waypoints.toMutableList().apply { add(bestIndex, point) }
    }
}

File: apps/ios/Sources/FeatureMap/RouteViewModel.swift
import Foundation
import Observation
import CoreModel
import CoreData

/// Drives route building: collect waypoints, solve (snap-to-trail via the
/// pathfinder, or straight legs), preview the polyline, and save the result as a
/// path. Mirrors `feature.map.RouteViewModel` (Android).
@MainActor
@Observable
public final class RouteViewModel {
    /// How the track's geometry is formed. One tool, three modes (mirrors
    /// Android's CreateTrack), folding the old separate "Measure" in — Line/Draw
    /// with the live distance readout *are* the measure tool.
    public enum Mode: String, CaseIterable, Sendable {
        case route   // tap waypoints; snap to trails via the solver
        case line    // tap waypoints; straight legs
        case draw    // freehand finger stroke
        public var label: String {
            switch self { case .route: "Route"; case .line: "Line"; case .draw: "Draw" }
        }
    }

    public var mode: Mode = .route
    public var preset: RoutePreset = .balanced
    public private(set) var waypoints: [LatLng] = []
    /// Freehand points captured in Draw mode (separate from tapped `waypoints`).
    public private(set) var drawPoints: [LatLng] = []
    /// The polyline to draw — the solver preview/result, or the straight legs.
    public private(set) var geometry: [LatLng] = []
    public private(set) var plan: RoutePlan?
    public private(set) var isSolving = false

    private let routeRepository: RouteRepository
    private let pathRepository: PathRepository
    private var solveTask: Task<Void, Never>?
    /// Snapshots of `waypoints` before each edit, for multi-level undo (capped).
    private var undoStack: [[LatLng]] = []
    private static let undoLimit = 20

    public var canUndo: Bool { !undoStack.isEmpty }

    public init(routeRepository: RouteRepository, pathRepository: PathRepository) {
        self.routeRepository = routeRepository
        self.pathRepository = pathRepository
    }

    public func addWaypoint(_ point: LatLng) {
        guard mode != .draw else { return }   // Draw uses the freehand stroke, not taps
        pushUndo(); waypoints.append(point); resolve()
    }

    /// Begin a fresh freehand stroke (drag start). Replaces any prior stroke.
    public func beginStroke() {
        guard mode == .draw else { return }
        pushUndo()
        drawPoints = []
    }

    /// Append a freehand point (drag move), throttled so we don't store hundreds
    /// of near-duplicate points per second.
    public func appendDrawPoint(_ point: LatLng) {
        guard mode == .draw else { return }
        if let last = drawPoints.last, GeoMetrics.haversineMeters(last, point) < 3 { return }
        drawPoints.append(point)
        resolve()
    }

    public func removeLast() {
        guard !waypoints.isEmpty else { return }
        pushUndo()
        waypoints.removeLast()
        resolve()
    }

    /// Remove a specific waypoint. A route needs ≥2 points, so dropping below that
    /// clears the plan but keeps the remaining points.
    public func removeWaypoint(at index: Int) {
        guard waypoints.indices.contains(index) else { return }
        pushUndo()
        waypoints.remove(at: index)
        resolve()
    }

    /// Reorder a waypoint (the editor's up/down controls).
    public func moveWaypoint(from: Int, to: Int) {
        guard waypoints.indices.contains(from), waypoints.indices.contains(to), from != to else { return }
        pushUndo()
        let point = waypoints.remove(at: from)
        waypoints.insert(point, at: to)
        resolve()
    }

    /// Reposition a waypoint (dragging its pin on the map).
    public func moveWaypoint(at index: Int, to point: LatLng) {
        guard waypoints.indices.contains(index) else { return }
        pushUndo()
        waypoints[index] = point
        resolve()
    }

    /// Insert a stop at the segment where it adds the least detour (Android's
    /// `addStop`). Falls back to appending when there's nothing to insert between.
    public func insertStop(_ point: LatLng) {
        pushUndo()
        waypoints = Self.insertLeastDetour(waypoints, point)
        resolve()
    }

    /// Revert the last edit.
    public func undo() {
        guard let previous = undoStack.popLast() else { return }
        waypoints = previous
        resolve()
    }

    public func clear() {
        solveTask?.cancel(); solveTask = nil
        waypoints = []; drawPoints = []; geometry = []; plan = nil; isSolving = false
        undoStack = []
    }

    private func pushUndo() {
        undoStack.append(waypoints)
        if undoStack.count > Self.undoLimit { undoStack.removeFirst() }
    }

    /// Insert `point` into `waypoints` at the position that minimises added detour.
    static func insertLeastDetour(_ waypoints: [LatLng], _ point: LatLng) -> [LatLng] {
        guard waypoints.count >= 2 else { return waypoints + [point] }
        var bestIndex = waypoints.count
        var bestDelta = Double.greatestFiniteMagnitude
        for i in 0..<(waypoints.count - 1) {
            let a = waypoints[i], b = waypoints[i + 1]
            let delta = GeoMetrics.haversineMeters(a, point) + GeoMetrics.haversineMeters(point, b)
                - GeoMetrics.haversineMeters(a, b)
            if delta < bestDelta { bestDelta = delta; bestIndex = i + 1 }
        }
        var result = waypoints
        result.insert(point, at: bestIndex)
        return result
    }

    public func setMode(_ newMode: Mode) {
        guard newMode != mode else { return }
        // Switching modes starts a fresh track (tapped waypoints and freehand
        // strokes don't mix), matching Android's openTrackTool reset.
        mode = newMode
        waypoints = []; drawPoints = []; geometry = []; plan = nil; undoStack = []
        solveTask?.cancel(); solveTask = nil; isSolving = false
    }
    public func setPreset(_ newPreset: RoutePreset) { preset = newPreset; if mode == .route { resolve() } }

    /// Persist the solved route as a recorded path.
    public func saveAsPath(name: String) {
        guard let plan, !plan.geometry.isEmpty else { return }
        let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
        let path = SavedPath(
            id: "route-\(UUID().uuidString)",
            name: trimmed.isEmpty ? "Route" : trimmed,
            path: GeoPath(points: plan.geometry, source: .route, distanceM: plan.distanceM, ascentM: plan.ascentM),
            activityKind: .hiking
        )
        Task { [pathRepository] in await pathRepository.upsert(path) }
    }

    // MARK: - Solve

    private func resolve() {
        solveTask?.cancel(); solveTask = nil
        let previousGeometry = plan?.geometry   // keep the last solved line while re-solving
        plan = nil

        if mode == .draw {
            geometry = drawPoints
            isSolving = false
            guard drawPoints.count >= 2 else { return }
            let distance = GeoMetrics.pathLengthMeters(drawPoints)
            plan = RoutePlan(distanceM: distance, durationS: distance / 1.3, ascentM: 0,
                             onTrailPct: 0, surfaces: [:], geometry: drawPoints)
            return
        }

        guard waypoints.count >= 2 else { geometry = waypoints; isSolving = false; return }

        switch mode {
        case .draw:
            break   // handled above
        case .line:
            geometry = waypoints
            let distance = GeoMetrics.pathLengthMeters(waypoints)
            plan = RoutePlan(distanceM: distance, durationS: distance / 1.3, ascentM: 0,
                             onTrailPct: 0, surfaces: [:], geometry: waypoints)
            isSolving = false
        case .route:
            // Hold the previous solved geometry (graceful re-solve) so the line
            // doesn't snap to straight segments mid-edit; fall back to straight.
            geometry = previousGeometry ?? waypoints
            isSolving = true
            let stream = routeRepository.planStream(points: waypoints, preset: preset, profile: "foot")
            solveTask = Task { [weak self] in
                for await event in stream {
                    switch event {
                    case .progress(let coords): self?.geometry = coords
                    case .result(let plan): self?.plan = plan; self?.geometry = plan.geometry; self?.isSolving = false
                    case .failure: self?.isSolving = false
                    }
                }
                self?.isSolving = false
            }
        }
    }
}
